### PR TITLE
virt: Detect WSL environment as a container (id: wsl)

### PR DESCRIPTION
**Pull request by rbalint**

### CODE REVIEW

**Comment by rbalint** [comment]
Many people would like to detect if their project is running under WSL and systemd could provide the standard detection mechanism.
https://github.com/Microsoft/WSL/issues/423

**Comment by poettering** [comment]
Hmm, is wsl a container mgr though? not sure what it is, but why would argue it qualifies to be considered a container mgr?

**Comment by poettering** [comment]
also, does systemd even run on wsl?

**Comment by rbalint** [comment]
WSL is not a classic container, but it is closest to the container type if we have to pick between a vm, a container or a chroot. It seems to fit the description here: https://www.redhat.com/en/topics/containers/whats-a-linux-container .

Systemd does not run on WSL as init yet, but there are developers interested in running it and I expect it be possible in the future. https://github.com/Microsoft/WSL/issues/994 . Right now /sbin/init is symlinked to systemd, but a statically linked binary is run as init: /init.

Handling WSL as a container would make units using ConditionVirtualization=!container do the right thing when systemd service files become supported in the WSL environment, but since we can argue why WSL is _not_ a container as we know them I'm also open to creating a new type.

Examples where ConditionVirtualization or systemd-detect-virt is used:
https://codesearch.debian.net/search?q=ConditionVirtualization&perpkg=1
https://codesearch.debian.net/search?q=systemd-detect-virt&perpkg=1

By a quick glance those would behave correctly in WSL with this code change.

**Comment by xnox** [comment]
It's closer to a docker thing than to a choort or a VM. It runs it's own init, and spawns new instances per-user / per-terminal etc. It executes unmodified ELF binaries. It cannot really run systemd (nor should), but it can run systemd-detect-virt.

**Comment by rbalint** [comment]
@poettering The immediate benefit of including this change in systemd is making package upgrades work properly when the upgrade scripts rely on systemd-detect-virt to skip operations failing in containers, such as updating grub:
https://bugs.launchpad.net/ubuntu/+source/grub2/+bug/1816753

File: man/systemd-detect-virt.xml [hunk 1]
```diff
@@ -160,6 +160,11 @@
             <entry><varname>rkt</varname></entry>
             <entry>rkt app container runtime</entry>
           </row>
+
+          <row>
+            <entry><varname>wsl</varname></entry>
+            <entry>Windows Subsystem for Linux</entry>
```
Comment (keszybz) [inline man/systemd-detect-virt.xml]: A `<ulink>` would be welcome here.

File: src/basic/virt.h [hunk 2]
```diff
@@ -32,6 +32,7 @@ enum {
         VIRTUALIZATION_OPENVZ,
```
Comment (keszybz) [inline src/basic/virt.h:32]: So we will now have a situation where "microsoft" is one thing, and "wsl" is another. That's rather confusing. The problem was introduced when we decided to use "microsoft" as the name of a virtualization mechanism. We should probably rename it to something more specific. But that's independent of this PR.

**Comment by keszybz** [comment]
Yeah, I think we can should merge this. Classifying WSL as a "container" is technically incorrect, but for practical purposes treating it as such seems appropriate.

**Comment by poettering** [comment]
hmm, so ok, let's merge this, but please extend the man page discussing briefly that this is not strictly a container, but something like a hybrid between a VM and a container or so, and for practical purposes considered a container.

In the long run something similar might pop up with ABI layers from other OSes I figure (let's say the linux api that freebsd has) maybe one day we want to have a third category, besides vms and containers called "kernel api layer" or so. but I figure we shouldn't bother with that now until anyone comes up with the actual patch for detecting bsd. And even if we have a third category like this I i figure all those cases should be considered containers still, after all systemd is never going to run as PID 1 of the main system on those layers (i.e. if windows itslef would be brought up by systemd, that'd be grant of course ;-)).

File: src/basic/virt.h (hunk 2 above)
Comment (rbalint) [reply src/basic/virt.h:32]: Done in a separate commit.

**Comment by rbalint** [comment]
@poettering Before I trigger all the tests again how about this wording?:

"<ulink url="https://docs.microsoft.com/en-us/windows/wsl/about">Windows Subsystem for Linux</ulink> which is not a Linux container, but an environment for running Linux userspace applications on top of the Windows kernel providing a Linux-compatible interface. For practical purposes it is handled as a container since multiple wsl environments share the same kernel and user space processes should generally behave like when being run in a container.</entry>"

**Comment by keszybz** [comment]
I assume that this text will go in as a separate paragraph after the table or a footnote.

Maybe something like this:
"Windows  Subsystem for Linux is not a Linux container, but an environment  for running Linux userspace applications on top of the Windows kernel using a Linux-compatible interface. WSL is categorized as a container for practical purposes. Multiple WSL environments share the same kernel and user space process space, and services should generally behave like when being run in a container."

**Comment by bjeanes** [comment]
My `/proc/sys/kernel/osrelease` in my Ubuntu WSL image has "microsoft" spelled lowercase so this check fails.

``` sh-session
$ cat /proc/sys/kernel/osrelease
4.19.84-microsoft-standard
```

Maybe this needs to be case-insensitive?

**Comment by bjeanes** [comment]
This was an image created under WSL1 and changed to WSL2 later (using `SetVersion`), if that matters.

**Comment by xnox** [comment]
@rbalint ping

@bjeanes 

I guess i should setup wsl2. Should wsl2 still be detected as container, or a VM?
Is there any difference between brand new wsl2 instance, versus one wsl2 one that got upgraded from wsl1?

**Comment by rbalint** [comment]
@xnox I asked @benhillis to provide a future-proof way of detecting WSL2 ~10 months ago: https://github.com/microsoft/WSL/issues/423#issuecomment-492301332
Until this is answered systemd should not add ad-hoc detection methods IMO, thus WSL2 detection is blocked on the answer from Microsoft's WSL team.

**Comment by bjeanes** [comment]
> Should wsl2 still be detected as container, or a VM?

I don't know. I'm new to WSL myself as I generally develop directly on a Unix machine but am giving it a red hot go.

> Is there any difference between brand new wsl2 instance, versus one wsl2 one that got upgraded from wsl1?

Also not sure on this, but I can't imagine the differences are substantial beyond the execution itself, as it appears you can freely change a WSL distro between WSL1 and WSL2.

I could try changing mine back to WSL1 and see if the `/proc/sys/kernel/osrelease` is different. It may simply be that the implementation in this PR (based on the guidance in https://github.com/microsoft/WSL/issues/423) only ever applied to that specific distro, at that specific time. There may not yet be a clear and unambiguous way to identify WSL via this method.

**Comment by bjeanes** [comment]
> Until this is answered systemd should not add ad-hoc detection methods IMO, thus WSL2 detection is blocked on the answer from Microsoft's WSL team.

I think this is very sensible. I just wanted to leave a comment to make it clear that while there was WSL detection merged, I don't think it actually works across the board (if at all).

**Comment by bjeanes** [comment]
@rbalint you may have noticed then, that I also left a comment on that microsoft/WSL#423 issue. I have the following env vars present:

``` sh-session
$ env | grep -i wsl
WSL_INTEROP=/run/WSL/18840_interop
WSL_DISTRO_NAME=Ubuntu
```

I haven't looked into whether these are available to all procesess or just from my interactive session, though.

**Comment by keszybz** [comment]
I think we should accept "microsoft" too. I opened a PR with that.

**Comment by bjeanes** [comment]
I hope they create an official way to check this, because my fear is that they'll ship a custom kernel for Azure that might also have `microsoft` in the kernel identifier and then this will trigger a false positive.

**Comment by bluca** [comment]
Issued a revert for "microsoft" in https://github.com/systemd/systemd/pull/15614 as we have cloud images that will show it too - WSL2 will appear in osversion soon-ish
